verifiers/internal/gcb/provenance.go

```diff
@@ -355,32 +356,56 @@ func (self *Provenance) VerifySourceURI(expectedSourceURI string, builderID util
 	if len(materials) == 0 {
 		return fmt.Errorf("%w: no materials", serrors.ErrorInvalidDssePayload)
 	}
-	uri := materials[0].URI
 
-	// It is possible that GCS builds at level 2 use GCS sources, prefixed by gs://.
-	if strings.HasPrefix(uri, "https://") && !strings.HasPrefix(expectedSourceURI, "https://") {
+	uri, err := url.Parse(materials[0].URI)
+	if err != nil {
+		return fmt.Errorf("parsing materials URI: %w", err)
+	}
+
+	expectedURI, err := url.Parse(expectedSourceURI)
+	if err != nil {
+		return fmt.Errorf("parsing materials URI: %w", err)
+	}
+
+	// If users do not specify a scheme, use HTTPS by default.
+	if !expectedURI.IsAbs() {
 		expectedSourceURI = "https://" + expectedSourceURI
+		expectedURI, err = url.Parse(expectedSourceURI)
+		if err != nil {
+			return fmt.Errorf("parsing materials URI: %w", err)
+		}
+	}
+
+	// Expect a direct match on scheme and hostnames.
+	if expectedURI.Scheme != uri.Scheme {
+		return fmt.Errorf("%w: expected schemes to match '%s', got '%s'",
+			serrors.ErrorMismatchSource, expectedURI.Scheme, uri.Scheme)
+	}
+	if uri.Host != expectedURI.Host {
+		return fmt.Errorf("%w: expected host match '%s', got '%s'",
+			serrors.ErrorMismatchSource, uri.Host, expectedURI.Host)
 	}
 
-	var err error
 	v := builderID.Version()
 	switch v {
 	case "v0.2":
-		// In v0.2, it uses format
-		// `https://github.com/laurentsimon/gcb-tests/commit/01ce393d04eb6df2a7b2b3e95d4126e687afb7ae`.
-		if !strings.HasPrefix(uri, expectedSourceURI+"/commit/") &&
-			!strings.HasPrefix(uri, expectedSourceURI+"#") {
-			return fmt.Errorf("%w: expected '%s', got '%s'",
-				serrors.ErrorMismatchSource, expectedSourceURI, uri)
+		// In v0.2, it uses the formats
+		// `https://github.com/laurentsimon/gcb-tests/commit/01ce393d04eb6df2a7b2b3e95d4126e687afb7ae`
+		// `gs://slsa-tooling_cloudbuild/source/1663616632.078353-fc7db143dcc64b5f9fe71d0497125ca1.tgz#1663616635134845`
+		if uri.Fragment == "" && !strings.HasPrefix(uri.Path, expectedURI.Path+"/commit/") {
+			return fmt.Errorf("%w: expected matching paths '%s', got '%s'",
+				serrors.ErrorMismatchSource, expectedSourceURI, uri.String())
+		}
+		if uri.Fragment != "" && !strings.HasPrefix(uri.Path, expectedURI.Path) {
```

Comment:
When comparing `gs://` URLs this check is maybe missing comparing the host/bucket name?

https://go.dev/play/p/hgDTpa5CX--

This applies to the other URL comparisons you are doing above and below as well.

Reply:
closing because this PR wont change logic